KEP-1645: Add specification for multi-network scenario (#3045)
kubernetes/enhancements: keps/sig-multicluster/1645-multi-cluster-services-api/README.md

  The cluster name should be consistent for the life of a cluster and its
  membership in the clusterset. Implementations should treat name mutation as
  a delete of the membership followed by recreation with the new name.

> - **cluster network** - An identifier for the cluster network. Each cluster can have an optional name that can identify the network its running in. The network name must be a valid [RFC

Reviewer: This makes an implicit assumption that egress network connectivity is mapped via identifiers to figure if given traffic needs to go through a proxy or is a direct connection. Has it been considered whether or not to express that specifically (e.g. the API is focused on "this traffic is proxied" vs saying "we have labeled our networks and this labeling implicitly implies that there is some proxying". The second statement assumes more about the environment and imposes a specific naming, labeling mechanism.

Author: Yes, will reword this to more generic after the right name is chosen in favor of `network`

The `ServiceImport.Spec.IP` (VIP) can be used to access this service from within
this cluster.

#### Multi-network scenario

> One or more clusters in a ClusterSet can be running on a discrete network (a non-flat network). An MCS controller can use the `network.k8s.io` `ClusterProperty` to determine if a cluster in a `ClusterSet` is running on a discrete network. Note that the endpoints of the `EndpointSlice` for a cluster on discrete network may only be representative of the pods backing the multi-cluster service and not the real pod addresses.

Reviewer 2: Thinking more about the definition here.  Maybe examples would help?

Consider a cluster with "native" pod IPs.  In that case, the pod network is the same as the node network.  Easy

Consider a cluster with "native" pod IPs, but the pod IPs are in a different subnet.  That's probably also the same as the node network (no visible routing?).

Consider 2 clusters with "native" pod IPs, on different networks where those networks are explicitly peered.  That's more or less the same as the previous case, but "feels" different.

Consider a cluster where the Pods exist on an overlay network (e.g. Flannel).  It's pretty clear that nodes and pods are NOT on the same network.  Which network goes into this property?

Consider a cluster where the Pods are "on" the node network (no encap) but use non-advertised IPs.  Which network goes into this property?

Reviewer 2: @caseydavenport since I know Calico / Canal / Flannel dances in this space.

Author: @thockin Thanks for writing down the scenarios, the `network` property refers to the `pod network`. We had some discussion last week and feedback was to change the property name from `network` to something else as `network` is too generic and overloaded. I will update the PR after the discussion, please let me know if you want to be in that meeting.

> One of more clusters in a ClusterSet can be running on a discrete network (a non-flat network). An MCS controller can use the `network.k8s.io` `ClusterProperty` to determine if a cluster in a `ClusterSet` is running on a discrete network. In this scenario. Note that the endpoints of the `EndpointSlice` for a cluster on discrete network may only be representative of the pods backing the multi-cluster service and not the real pod addresses.

Reviewer 3: ```suggestion
One or more clusters in a ClusterSet can be running on a discrete network (a non-flat network). An MCS controller can use the `network.k8s.io` `ClusterProperty` to determine if a cluster in a `ClusterSet` is running on a discrete network. Note that the endpoints of the `EndpointSlice` for a cluster on discrete network may only be representative of the pods backing the multi-cluster service and not the real pod addresses.
```

##### Known limitation

> In a multi-network scenario where the `EndpointSlice`s do not contain the actual pod addresses, there isn't currently a way (K8s native support) to proportionately distribute the traffic based on the actual number of pods. There is active ongoing work in SIG-Network to add an attribute to represent the number of endpoints for `EndpointSlice`s. This will provide a way for kube-proxy to load balance across `EndpointSlice`s .

Reviewer 2: I don't think there's actually a KEP for "weight" in EndpointSlices, is there?  @robscott @JeremyOT @lauralorenz @bowei 

Given the KEP freeze next week, there's still time, but not very much, for v1.24, if someone jumps on this.  Otherwise the earliest possible alpha is 1.25 (summer)

Author: @thockin Let me start a draft and share by tomorrow

Reviewer 4: There is not a KEP for weight. The primary concern was that of scalability. EndpointSlices are already updated very frequently. Depending on the implementation, weighting could exponentially increase those updates.

/cc @wojtek-t

Reviewer 2: we have to have it SOMEHOW, right?  I don't see how this can go very far without ...

We might need some more clever design for the API, but we need ... something?

Reviewer 4: It _may_ be a reasonable compromise to have per-slice weight instead of per-endpoint weight. That seems like it could potentially work for this use case while still reasonably limiting writes. For topology aware routing we would likely want different weights depending on where the request was originating from. With that said, per-slice weight is a poor solution when there are a small number of endpoints as it would dramatically increase the number of slices.

Author: @robscott Great points, we can discuss more about the tradeoffs over the meeting next week.

Reviewer 2: If we ever want to allow multi-network MCS, we need to revisit this.

The `ServiceImport.Spec.IP` (VIP) can be used to access this service from within
this cluster.

#### Multi-network scenario

> One of more clusters in a ClusterSet can be running on a discrete network (a non-flat network). An MCS controller can use the `network.k8s.io` `ClusterProperty` to determine if a cluster in a `ClusterSet` is running on a discrete network. In this scenario, the ServiceImport and the EndpointSlice can be generated to point to the Gateway/Loadbalancer of destination cluster. Note that the endpoints of the `EndpointSlice`s may only be representative of the pods backing the multi-cluster service and not real pod addresses, and may actually represent multiple pods. The MCS controller implementation is responsible for determining the Gateway/Loadbalancer in a cluster for a non-flat network. An MCS controller implementation can use a K8s [Gateway](https://gateway-api.sigs.k8s.io/) defined with an annotation `network.k8s.io: mcs-gateway` to match against a ServiceExport with the same annotation to determine the Gateway to be used. 

Reviewer 2: I don't understand the point about a gateway annotation?

Reviewer 5: I'm not sure we need to call out the gateway annotation. In a previous comment I suggested we trim to just recommending that a standard cluster property (`network.k8s.io` or other) is used for the cluster's network. Really, that's more to help fix the lack of cluster self-awareness than to really enable implementations. If there's other work going on there it would be best not to duplicate. @thockin how can I get plugged in? multi-net and multi-cluster have a lot of overlap.

I believe this spec already provides the flexibility required for an implementation to do whatever it needs to for multi network - if we need to more explicitly highlight that the IPs in EndpointSlices don't have to be real IPs from another cluster we should start there.

Author: @JeremyOT I removed the gateway references, can I keep the PR until we find the details about the ongoing effort?
@thockin Thanks for the feedback, would love to learn more about the ongoing effort.

Author: @thockin Can you please let me know about the ongoing effort?

Reviewer 2: multi-net exists in the form of `multus` and there have been some new mumbles about making it more first-class.  I think it's probably safe  to assume that there's a "default' network or "pod network", and this property models that?

Author: Yep, thats right.

The `ServiceImport.Spec.IP` (VIP) can be used to access this service from within
this cluster.

#### Multi-network scenario

> One of more clusters in a ClusterSet can be running on a discrete network (a non-flat network). An MCS controller can use the `network.k8s.io` `ClusterProperty` to determine if a cluster in a `ClusterSet` is running on a discrete network. In this scenario, the generated EndpointSlice in an importing cluster will point to the FQDN (Fully Qualified Domain Name) of the Gateway/Loadbalancer of the cluster this `EndpointSlice` belongs to. The MCS controller implementation is responsible for determining the Gateway/Loadbalancer in a cluster for a non-flat network. A sample implementation can use a K8s [Gateway](https://gateway-api.sigs.k8s.io/) defined with an annotation `network.k8s.io: mcs-gateway` to match against a ServiceExport with the same annotation to determine the FQDN associated with it. This FQDN would then be used in the EndpointSlice generated for this cluster in the importing clusters. The MCS controller may implement creating the K8s Gateway and the necessary routing rules to automate the configuration.

Reviewer 5: I think we need to relax this quite a bit. In multi-network, we do want to suggest that the `network.k8s.io` `ClusterProperty` is used, but that's probably where our spec ends.

We can use this as a place to reiterate that the endpoints in an `EndpointSlice` may only be representative of the pods backing the multi-cluster service and not real pod addresses, and may actually represent multiple pods (e.g. when a gateway is used).

Beyond that, we tend to favor flexibility. FQDNs addresses in `EndpointSlice`s cannot realistically be part of the spec because they don't actually do anything out-of-the-box. It might be better instead to have the implementation set the `ServceImport`'s `ClusterSetIP` to a gateway address, but that should probably be left to the implementation.

The main point we want to cover here is that so long as the rest of the spec is met, it doesn't matter how you wire up the backends.

Author: Great feedback thanks!

The `ServiceImport.Spec.IP` (VIP) can be used to access this service from within
this cluster.

#### Multi-network scenario

> One or more clusters in a ClusterSet can be running on a discrete network (a non-flat network). An MCS controller can use the `network.k8s.io` `ClusterProperty` to determine if a cluster in a `ClusterSet` is running on a discrete network. Note that the endpoints of the `EndpointSlice` for a cluster on discrete network may only be representative of the pods backing the multi-cluster service and not the real pod addresses.

Reviewer 6: > Note that the endpoints of the `EndpointSlice` for a cluster on discrete network may only be representative of the pods backing the multi-cluster service

How does the MCS controller know how to populate the endpoint slices with the appropriate endpoints? Does it use the ClusterProperty's `network` value? Or is that a network name that can signal to the MCS controller to use some mapping of endpoints that exists $somewhere?

Some common API here may be useful, ie: an annotation on either the `ServiceImport`, or the `ServiceExport` would allow for different services to leverage different services (may be for some QOS isolation reasons). 

The annotation could target an Ingress endpoint, public hostname, IP address, or in-cluster service.

`proxy.mcs.k8s.io: <prox-url>`

I understand that the MCS controller implementation is not part of the multi cluster API, but it is at the very least useful to think of a full end to end solution to ensure the API is robust.
